bgpv2: decouple instance registration and reconcile errors

harsimran-pabla opened the pull request:
BGP instances are independent of each other, if there is failure to register or reconcile one instance, we should not abort registering or reconciling remaining instances. We calculate diff for registering and reconciling on each trigger, so failed instances will be retried again when next trigger is received.

rastislavs commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -816,22 +816,18 @@ func (m *BGPRouterManager) ReconcileInstances(ctx context.Context,
-		if err := m.reconcileV2(ctx, rd); err != nil {
-			return err
-		}
+		m.reconcileV2(ctx, rd)
```
This change would mean that `BGPRouterManager.ReconcileInstances()` will never return an error upon reconcile, right?

That makes the [retry logic in the Controller](https://github.com/cilium/cilium/blob/e53c222f3bc2517a35fc6f3db0df81cbb27befc4/pkg/bgpv1/agent/controller.go#L210-L237) ineffective (as `BGPRouterManager.ReconcileInstances()` is called from the Controller), so we should either move the retry logic, or return an error from `ReconcileInstances` if any of the `registerV2` / `reconcileV2` attempts failed.

harsimran-pabla replied:
Yes, you're right! So, this time combining all the errors and bubbling up, so retry logic can kick in.

harsimran-pabla committed:
bgpv2: decouple instance registration and reconcile errors

BGP instances are independent of each other, if there is failure to
register or reconcile one instance, we should not abort registering or
reconciling remaining instances.

On failure, errors are combined and bubbled up for retry.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

rastislavs reviewed (approved):
LGTM now!
